### PR TITLE
New demo scene using Unity Terrain

### PR DESCRIPTION
**Pull request by williamtsoi**
Hi Reese,
Thanks for making this open source repo! 

I'm working on an ECS project using Unity Terrain, and wanted to create a quick demo with this to get pathfinding working. but I seem to be running into some performance issues.

Here is the original performance demo, spawned 2k ish agents and I'm getting 155fps (they all pathed away by the time i took the screenshot but they did spawn!)

And here is the terrain demo scene, where it is struggling at 25fps for 200ish entities.

Here's the profiler, but I'm not sure where to look or what this really means though.

Any chance you might be able to have a look or point me in the right direction for this?

Thanks in advance :)

### CODE REVIEW

**Comment by williamtsoi** [comment]
Thanks for your pointer towards the invalid destination issue. Looks like that was definitely at least part of the issue.

I've added a check using a raycast for the randomly generated destination to make sure it lands onto the terrain collider before adding it as a `NavNeedsDestination` component and that seems to have helped. 
Also added a check for non successful `NavPlanning`s and to remove the `NavPlanning` tag so it will look for a new one, is this what you mean for #35 ?

Also do let me know if you'd like me to comment/commit in a different style etc, first time collaborating on a project, and not sure if there is a standard to how how things are done.

File: Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs
Comment (williamtsoi) [inline Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs]: Not sure when to us e CollisionWorld vs PhysicsWorld, but the [docs examples](https://docs.unity3d.com/Packages/com.unity.physics@0.4/manual/collision_queries.html) are in CollisionWorld so thats what I've used here, but noticed the rest of your project was using PhysicsWorld, so I can change this to be consistent if you would prefer.

Comment (reeseschultz) [reply Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs]: Functionally I'm not sure what the difference is between the two, and have been wondering about that as well. I may have to dig into Unity's code to find out. One may be more performant than the other, but I'm not sure which. For consistency with this project, you can just use the physics world, but if it turns out using the collision world is objectively better somehow, then we can always refactor everything in one fell swoop later.

**Comment by reeseschultz** [comment]
For #35, you are correct that the `NavPlanning` tag ought to be removed upon failure. `NavNeedsDestination` may need removal too, depending on how you want to handle things conditionally to get a new destination. What I do intend to do with #35 later is passthrough different failure statuses, so the user knows how the planning went in detail. For example, a user who receives an `OutOfNodes` error may want to leverage a procedure that adds nodes in that case. `WrongMagic` might imply that the nav mesh needs re-baking, possibly. Multiple flags can be set, complicating matters, so they may just have to be passed through as bit flags in a new component. I think the change you already made should suffice for now, and will partially account for #35. For now we can just assume if `NavPlanning` component is removed, but `NavNeedsDestination` exists and there is no `NavLerping`, something must have went wrong.

If you want to handle that more semantically, feel free to add code as you see fit. Your reasoning looks great to me so far. In terms of style, by the way, don't worry much about that either. I can make edits when the PR is done-ish, and then we can squash everything upon merge. We can also figure out how you might want credit for your changes. Also, thank you! I really appreciate having help on this project.

**Comment by williamtsoi** [comment]
Thanks :) cleaned up a bit based on your comments, also renamed it to `GetPointOnSurfaceLayer` to make it more explicit. Looks like Extension methods only work with `this` so kept it as that, but moved it into `NavUtil`. Not done much else other than that yet.

Its its now getting similar performance to the performance demo, which I'm happy about. I think next step for me is to see about getting the agents aware of the terrain height so they don't sync into the ground (or fly? not sure about that one on the right). I'm thinking adding a collider to the agents and use that to set the `y` to see if that will still run smoothly.
2020-07-07_LI_Moment_Moment

Also currently I am actually manually creating the `RenderBounds` for the terrain in order for the spawner to pick random locations, ideally I need to figure out how to actually get a valid render bounds for a terrain, but I think this is lower priority since the spawning isn't directly related to the pathing, I guess another method can be used for picking locations if it comes down to that.

File: Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs
Comment (williamtsoi) [reply Assets/Scripts/Nav/Destination/NavTerrainDestinationSystem.cs]: Agreed - updated.

**Review by reeseschultz — CHANGES_REQUESTED** [review]
I'm not confident that this approach (referring to the `GroundedSystem`) will scale well, especially not when there are a lot of parent transformations taking place. This is too much raycasting, which we should avoid. If agents aren't grounded, we need to reason about *why* that is rather than bruteforcing a grounding strategy. Most of the agents seem to behave well most of the time, so there are probably no more than a couple corner cases resulting in outcomes that prevent grounding. We just need to figure out what those are and deal with them specifically.

My first instinct is to check how obstacles are detected in the `NavInterpolationSystem`. Before it was tweaked to its current state, I recall it causing some odd behaviour. It raycasts too, but the raycasting is significantly amortized to prevent scaling issues. I'm pretty sure if I set the `OBSTACLE_RAYCAST_DISTANCE_MAX` too low, I would observe that flying behavior in the `NavPerformanceDemo`. Obstacle detection is a bit of a hack, so I wouldn't be surprised if it was the culprit causing the behavior you're trying to avoid.

If we cannot determine a root cause, then an amortized approach could work, such as grounding only *once* when a new path is established or something, but I'd really prefer to avoid that too.

**Comment by williamtsoi** [comment]
Hi Reese,

Totally agree that Raycasting every frame shouldn't be the way to go. Was just surprised that it was actually running smoothly even with 1000k+ agents.

I was looking in the `NavInterpolationSystem` but it looks like the the obstacle avoidance is only related to jumping between surfaces so that doesn't seem to be able to either help me or cause me issues regarding this.

It seems that the cause is related to how the paths are generated between points on the Navmesh.
The path is a straight lines between points on the mesh, and this could mean that the straight line actually cuts through the terrain since the height between the two points can vary by a lot. The `NavInterpolationSystem` then uses the way point to move forward therefor also following that same line and cutting through terrain.
So I don't think there is an actual bug here per say, just that it cannot handle uneven terrain as is.
Because of this, I also don't think casting only at the start of path will fix it either, the problem will only manifest itself mid path and actually would be fine at start and end of paths.

Hope that makes sense.
image
image

What do you think of using a Collider for this instead of the Raycasting approach, or prehaps a throttled Raycast? 
I think the standard Unity agent uses a simplified Raycast as well for following the terrain, guessing by looking at this method name: `public bool SamplePathPosition(int areaMask, float maxDistance, out NavMeshHit hit);` [unity-decompiled/../NavMeshAgent.cs](https://github.com/jamesjlinden/unity-decompiled/blob/master/UnityEngine/UnityEngine/NavMeshAgent.cs)

I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't? Depending on the project, NavAgents might or might not already have an existing collider.

Let me know what you think, Cheers

**Comment by reeseschultz** [comment]
You raise good points, @williamtsoi. Thanks for bearing with me.

> I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't?

I like that a lot. I think adding a `NavTerrainCapable` component tag to `NavAgentStatus.cs` would do the trick. In the `NavGroundedAgentSystem`, it can query only agents that have said tag. You can define the tag as follows:

```csharp
/// <summary>Exists if the agent needs to traverse complex terrain.</summary>
public struct NavTerrainCapable : IComponentData { }
```

When spawning agents for the terrain demo, just add that tag to them and you should be good to go.

Beyond that, I don't think any further work would be required for this PR, unless there's more you feel needs doing. I don't think the optimizations are necessary for us to merge this PR if we leverage that component tag. Having it alleviates my concerns since then I don't have to worry about adverse effects on the other demos/use cases.

If interested, you could submit a subsequent PR later to potentially optimize the grounding for terrain. I like the idea of separate jobs for `NavTerrainCapable` agents that have `PhysicsCollider` components and those that do not, but the work involved may count as a feature in and of itself. And speculative throttled raycasting may or may not be yet another PR. We would want to confirm that the work really had a positive performance impact anyway. You would be under no obligation to submit any further PRs, but you're more than welcome to do so. I think the work you're doing here is fantastic.

**Comment by williamtsoi** [comment]
>I think maybe an optional tag that could be added a NavAgent entity which would allow it to use an existing collider on the entity if it already contains one, or Raycast as a fall back if it doesn't?

So after some investigation, looks like using a collider wont work how I imagined it could, the collision [Trigger Event](https://docs.unity3d.com/Packages/com.unity.physics@0.4/api/Unity.Physics.TriggerEvent.html), doesn't actually return the collision point, and the doing a collider cast would be more expensive than the ray cast, so I've removed that implementation and returned it to only using a Raycast again. I've added the optional tag for the `NavGroundedAgentSystem` and updated the spawner so it's isolated to the the terrain demo.

So the last thing I've I've been trying to add was to make the agent align the the surface normal - think cars, tanks etc.
I got close over the last few days, but I can't work out how to make it work correctly.
The current solution rotates it along the Z axis correctly, but not the X axis, so the agent is rotated sideways correctly, but not looking down the hill like it should.
The green line is the surface normal, the drawn blue line is where it should be looking, but it is instead facing forward on the blue+white line.
image

I've given it a good attempt but I don't think I'll be able to work this one out at the moment, would you rather I pulled this change out so we can complete this PR, or would you be able to maybe have a look or point me in the right direction?

**Comment by reeseschultz** [comment]
@williamtsoi I got you fam.

Getting the rotation right is a multi-step process. In other words, a pain in the ass. Here's one way to do it:

1. In the `NavAgent`, add this member:

```csharp
/// <summary>Normal of point below the agent.</summary>
public float3 SurfacePointNormal;
```

2. In the `NavAgentGroundedSystem`, now all you need is this:

```csharp
if (physicsWorld.CastRay(rayInput, out RaycastHit hit)) agent.SurfacePointNormal = hit.SurfaceNormal;
```

(But you're welcome to leave your debug code in place, by the way, I think it's useful. )

3. We'll keep all our interpolation in the `NavInterpolationSystem` not just because of the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), but to get the rotation right, so the end of the `NavWalkJob` should be replaced with this:

```csharp
translation.Value = Vector3.MoveTowards(translation.Value, localWaypoint, agent.TranslationSpeed * deltaSeconds);

var lookAt = NavUtil.MultiplyPoint3x4( // To world (from local in terms of destination surface).
    localToWorldFromEntity[agent.DestinationSurface].Value,
    localWaypoint
);

lookAt = NavUtil.MultiplyPoint3x4( // To local (in terms of agent's current surface).
    math.inverse(localToWorldFromEntity[surface.Value].Value),
    lookAt
);

lookAt.y = translation.Value.y;

var lookRotation = quaternion.LookRotationSafe(lookAt - translation.Value, math.up());

if (((Vector3)agent.SurfacePointNormal).magnitude > 0.1f)
    lookRotation = Quaternion.FromToRotation(localToWorldFromEntity[entity].Up, agent.SurfacePointNormal) * lookRotation;

rotation.Value = math.slerp(rotation.Value, lookRotation, deltaSeconds / 0.3f);
```
We're using both `quaternion` and `Quaternion` with a capital Q since for some reason there isn't feature parity between the two. To get around that I was going to use `Unity.Physics.Math.FromToRotation` for calculating the normal rotation, but it doesn't work as I expect, so we can just live with what amounts to an annoying style discrepancy for the sake of getting this working for now.

You'll note two magic numbers. The one used in the `SurfacePointNormal` magnitude check (`0.1f`) is probably fine. Although, you may want to add a rotation speed member called `RotationSpeed` to the `NavAgent` for the `slerp` operation, replacing the `0.3f` with it.

Other things I want to bring up:
* I think we should change the name of `NavGroundedAgentSystem` to `NavGroundingSystem` for consistency with the other systems (using present as opposed to past tense).
* I updated a bunch of stuff in the main branch, but I can take care of merging if you're not comfortable with that.

People will dig this feature/demo. Honestly they'll probably find this more useful than the transformation hierarchy stuff that I obsess over. Thanks for sticking with this.

**Comment by williamtsoi** [comment]
Thanks @reeseschultz , that helps a lot!

Didn't think of using the standard Quaternion for help.
Just had to tweak this line slightly to make it rotate fully to the normal `lookRotation = Quaternion.FromToRotation(math.up(), agent.SurfacePointNormal) * lookRotation;`.

Just wanted to do some final tests with the scene before signing this off so I added more unpathable areas to the terrain. 
Unfortunately I've noticed that the agents seems to be getting stuck along some of the edges and don't seem to get close enough to the `localWaypoint` for it to move onto the next waypoint. -at least that's what I think is happening from the initial investigation. I'll be digging into this some more. 

Hoping to get this hopefully-final issue resolved soon :)
Cheers

File: Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs
Comment (reeseschultz) [inline Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs]: That's a lot of flags for debugging. Might it be possible to reduce them, preferably down to one?

File: Assets/Scripts/Nav/Spawner/NavTerrainSpawner.cs
Comment (reeseschultz) [inline Assets/Scripts/Nav/Spawner/NavTerrainSpawner.cs:82]: Would you mind updating the `RotationSpeed` to `0.3f` in all the other nav-related spawners as well?

File: Packages/com.reese.nav/Agent/NavGroundingSystem.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/Agent/NavGroundingSystem.cs:60]: That looks much better!

File: Packages/com.reese.nav/Agent/NavInterpolationSystem.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/Agent/NavInterpolationSystem.cs:110]: Good! I didn't know about `math.length`.

File: Packages/com.reese.nav/Agent/NavPlanSystem.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/Agent/NavPlanSystem.cs:77]: Looks like there's an unintended extra newline here. Also, I'm curious, were all the line endings in this file changed?

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs:39]: Please replace "Extention" with "Extension."

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs]: Please replace "using raycast" with "by raycasting."

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs]: Please replace `position + -localToWorld.Up` with `position - localToWorld.Up`.

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs]: Please replace "was successful" with "is successful."

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs]: Please replace this and the next line (42 and 43) with `/// a position returned via out.`

File: Packages/com.reese.nav/NavUtil.cs
Comment (reeseschultz) [inline Packages/com.reese.nav/NavUtil.cs]: You can remove this line and the comma on the line above (53) since the `GroupIndex` default is equivalent.

**Review by reeseschultz — CHANGES_REQUESTED** [review]
@williamtsoi your code is looking great. I diligently reviewed it because you're coming very close to completion. I left a lot of comments about little stuff. It's absolutely not to be nit-picky, it's just stuff that I'd like addressed before we merge.

> Just wanted to do some final tests with the scene before signing this off so I added more unpathable areas to the terrain.
Unfortunately I've noticed that the agents seems to be getting stuck along some of the edges and don't seem to get close enough to the localWaypoint for it to move onto the next waypoint. -at least that's what I think is happening from the initial investigation. I'll be digging into this some more.

Please let me know if you end up needing help investigating that issue.

And thanks for the update! FYI, I created [a Discord channel](https://discord.gg/dQkjpKT) to centralize real-time discussion of my open source projects and stuff. If that helps/interests you in some way, you're welcome to join.

File: Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs
Comment (williamtsoi) [reply Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs]: Good point, updated and kept the toggles separate in the inspector, but clasped down to 1 flag per system that is being debugged.

File: Packages/com.reese.nav/Agent/NavPlanSystem.cs
Comment (williamtsoi) [reply Packages/com.reese.nav/Agent/NavPlanSystem.cs:77]: I'm not sure about the line ending, looks like it is currently CRLF just like the other files, not sure why it is flagging it whole file as modified here.

File: Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs
Comment (williamtsoi) [inline Assets/Scripts/Nav/NavTerrainDemo/DebugMode.cs:81]: Didn't realise I could use Debug.DrawLine from a System as well now, so moved this into a SystemBase Entities loop.

File: Packages/com.reese.nav/Agent/NavGroundingSystem.cs
Comment (williamtsoi) [inline Packages/com.reese.nav/Agent/NavGroundingSystem.cs]: Should not be setting the Agent's physical position by the offset. I think I misunderstood what the offset was for.

File: Packages/com.reese.nav/Agent/NavInterpolationSystem.cs
Comment (williamtsoi) [inline Packages/com.reese.nav/Agent/NavInterpolationSystem.cs]: Added offset when checking for distance to current waypoint, same with the checks in the rest of the file.

Looks like this was 1 of the issue which was causing the agent to stop mid path as the vertical offset was too great from the path's y position.

File: Packages/com.reese.nav/NavConstants.cs
Comment (williamtsoi) [inline Packages/com.reese.nav/NavConstants.cs:45]: Looks like this was was the cause of the other issue I was having.
When the nav mesh was made more complex due to the terrain having more unpathable areas, it was running out of nodes to find a valid path.

**Review by williamtsoi — COMMENTED** [review]
looks like adding the agent offset to the `NavInterpolationSystem` fixes the problem for the "Grounded" agents, but breaks the Agents for all the other demos.
So looking further into the offset on the NavAgent, it seems like the offset is also used when navigating between way points and updating the translation causing the Agent to float above the surface by the offset as well.
Is this use of the offset intended? Or should the offset only be used by the RayCasting functions so that it does not start below surface e.g. with finding the parent when spawning.

**Comment by williamtsoi** [comment]
Hi @reeseschultz , sorry its been a while since I was able to have a look at this. I've finally been able to resolve that final issue I was having with agents getting stuck on way points. 
Think this might be ready for the final review!

**Review by reeseschultz — APPROVED** [review]

**Comment by reeseschultz** [comment]
Thank you so much for your effort @williamtsoi, you did a great job. I'm merging this into a temporary feature branch to make some minor adjustments, then it will be merged into the main branch afterward.

**Pull request merged**
